src/llama-model.cpp

```diff
@@ -4574,6 +4624,192 @@ struct llm_build_llama : public llm_graph_context {
+            }
+
+            // For Granite architecture
+            if (hparams.f_residual_scale) {
+                cur = ggml_scale(ctx0, cur, hparams.f_residual_scale);
+            }
+
+            ggml_tensor * ffn_inp = ggml_add(ctx0, cur, inpSA);
+            cb(ffn_inp, "ffn_inp", il);
+
+            // feed-forward network
+            if (model.layers[il].ffn_gate_inp == nullptr) {
+
+                cur = build_norm(ffn_inp,
+                        model.layers[il].ffn_norm, NULL,
+                        LLM_NORM_RMS, il);
+                cb(cur, "ffn_norm", il);
+
+                cur = build_ffn(cur,
+                        model.layers[il].ffn_up,   model.layers[il].ffn_up_b,   NULL,
+                        model.layers[il].ffn_gate, model.layers[il].ffn_gate_b, NULL,
+                        model.layers[il].ffn_down, model.layers[il].ffn_down_b, NULL,
+                        NULL,
+                        LLM_FFN_SILU, LLM_FFN_PAR, il);
+                cb(cur, "ffn_out", il);
+            }
+
+            cur = ggml_add(ctx0, cur, ffn_inp);
+            cb(cur, "ffn_out", il);
+
+            cur = build_cvec(cur, il);
+            cb(cur, "l_out", il);
+
+            // input for next layer
+            inpL = cur;
+        }
+
+        cur = inpL;
+
+        cur = build_norm(cur,
+                model.output_norm, NULL,
+                LLM_NORM_RMS, -1);
+
+        cb(cur, "result_norm", -1);
+        res->t_embd = cur;
+
+        if (model.csm_output_cbook) {
+            // Sesame csm backbone
+            // hack: because n_cbook < n_vocab, we use the first logits for the output
+            int64_t n_vocab = model.tok_embd->ne[1];
+            int64_t n_codes = model.csm_output_cbook->ne[1];
+            ggml_tensor * last_h = cur;
+            cur = build_lora_mm(model.csm_output_cbook, cur);
+            cur = ggml_pad(ctx0, cur, n_vocab - n_codes, 0, 0, 0);
+
+            // project to csm decoder dim
+            last_h = build_lora_mm(model.csm_proj, last_h);
+            cb(last_h, "output_csm_proj", -1); // use callback to retrieve the result
+            ggml_build_forward_expand(gf, last_h);
+
+        } else if (model.csm_output_audio && ggml_nelements(cur)) {
+            // Sesame csm decoder
+            // hack: because n_audio < n_vocab, we use the first logits for the output
+            cur = build_lora_mm_id(model.csm_output_audio, cur, inp_pos);
+            int64_t n_vocab = model.tok_embd->ne[1];
+            int64_t n_codes = cur->ne[0];
+            cur = ggml_pad(ctx0, cur, n_vocab - n_codes, cur->ne[1], 0, 0);
+
+            // also get audio embeddings, which will be passed back to backbone to keep track of generation progress
+            if (ubatch.token) {
+                cb(input_embd, "output_audio_embd", -1);
+                ggml_build_forward_expand(gf, input_embd);
+            }
+
+        } else {
+            // otherwise, dummy output
```

Comment:
Can we reach this branch?

Reply:
I'm not 100% why but the warmup call runs this will a `ggml_nelements(cur) == 0`, which will trigger this branch. I assume that's because there is no output token being set in the batch